// ts-simple-ast/dist-scripts/scripts/inspectors/InspectorFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("../common");
var WrapperFactory_1 = require("./WrapperFactory");
var TsSimpleAstInspector_1 = require("./TsSimpleAstInspector");
var TsInspector_1 = require("./TsInspector");
var InspectorFactory = /** @class */ (function () {
    function InspectorFactory() {
        var wrapperFactory = new WrapperFactory_1.WrapperFactory();
        this.project = common_1.getProject();
        this.tsSimpleAstInspector = new TsSimpleAstInspector_1.TsSimpleAstInspector(wrapperFactory, this.project);
        this.tsInspector = new TsInspector_1.TsInspector(wrapperFactory, this.project);
    }
    InspectorFactory.prototype.getTsSimpleAstInspector = function () {
        return this.tsSimpleAstInspector;
    };
    InspectorFactory.prototype.getTsInspector = function () {
        return this.tsInspector;
    };
    InspectorFactory.prototype.getProject = function () {
        return this.project;
    };
    return InspectorFactory;
}());
exports.InspectorFactory = InspectorFactory;
