// ts-simple-ast/dist-scripts/scripts/inspectors/WrapperFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../src/utils");
var tsSimpleAst_1 = require("./tsSimpleAst");
var ts_1 = require("./ts");
var WrapperFactory = /** @class */ (function () {
    function WrapperFactory() {
        this.wrapperNodeCache = new utils_1.KeyValueCache();
        this.structureNodeCache = new utils_1.KeyValueCache();
        this.mixinNodeCache = new utils_1.KeyValueCache();
        this.tsNodeCache = new utils_1.KeyValueCache();
        this.tsNodePropertyCache = new utils_1.KeyValueCache();
    }
    WrapperFactory.prototype.getWrapperNode = function (classDeclaration) {
        var _this = this;
        return this.wrapperNodeCache.getOrCreate(classDeclaration, function () { return new tsSimpleAst_1.WrappedNode(_this, classDeclaration); });
    };
    WrapperFactory.prototype.getMixin = function (interfaceDeclaration) {
        var _this = this;
        return this.mixinNodeCache.getOrCreate(interfaceDeclaration, function () { return new tsSimpleAst_1.Mixin(_this, interfaceDeclaration); });
    };
    WrapperFactory.prototype.getStructure = function (interfaceDeclaration) {
        var _this = this;
        return this.structureNodeCache.getOrCreate(interfaceDeclaration, function () { return new tsSimpleAst_1.Structure(_this, interfaceDeclaration); });
    };
    WrapperFactory.prototype.getTsNode = function (interfaceDeclaration) {
        var _this = this;
        return this.tsNodeCache.getOrCreate(interfaceDeclaration, function () { return new ts_1.TsNode(_this, interfaceDeclaration); });
    };
    WrapperFactory.prototype.getTsNodeProperty = function (propertySignature) {
        var _this = this;
        return this.tsNodePropertyCache.getOrCreate(propertySignature, function () { return new ts_1.TsNodeProperty(_this, propertySignature); });
    };
    return WrapperFactory;
}());
exports.WrapperFactory = WrapperFactory;
