// tsuml/dist/core/templates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.templates = {
    composition: "+->",
    implementsOrExtends: function (abstraction, implementation) {
        return ("" + exports.templates.plainClassOrInterface(abstraction) +
            ("^-" + exports.templates.plainClassOrInterface(implementation)));
    },
    plainClassOrInterface: function (name) { return "[" + name + "]"; },
    colorClass: function (name) { return "[" + name + "{bg:skyblue}]"; },
    colorInterface: function (name) { return "[" + name + "{bg:palegreen}]"; },
    class: function (name, props, methods) {
        var pTemplate = function (property) { return property.name + ";"; };
        var mTemplate = function (method) { return method.name + "();"; };
        return ("" + exports.templates.colorClass(name) +
            ("[" + name + "|" + props.map(pTemplate).join("") + "|" + methods.map(mTemplate).join("") + "]"));
    },
    interface: function (name, props, methods) {
        var pTemplate = function (property) { return property.name + ";"; };
        var mTemplate = function (method) { return method.name + "();"; };
        return ("" + exports.templates.colorInterface(name) +
            ("[" + name + "|" + props.map(pTemplate).join("") + "|" + methods.map(mTemplate).join("") + "]"));
    }
};
