// ts-simple-ast/dist/compiler/aliases.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var typescript_1 = require("../typescript");
/* istanbul ignore next */
function propertyNameAliasValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.Identifier:
        case typescript_1.SyntaxKind.StringLiteral:
        case typescript_1.SyntaxKind.NumericLiteral:
        case typescript_1.SyntaxKind.ComputedPropertyName:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function accessorDeclarationAliasValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.SetAccessor:
        case typescript_1.SyntaxKind.GetAccessor:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function entityNameValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.Identifier:
        case typescript_1.SyntaxKind.QualifiedName:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function jsxChildValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.JsxText:
        case typescript_1.SyntaxKind.JsxExpression:
        case typescript_1.SyntaxKind.JsxElement:
        case typescript_1.SyntaxKind.JsxSelfClosingElement:
        case typescript_1.SyntaxKind.JsxFragment:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function jsxAttributeValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.JsxAttribute:
        case typescript_1.SyntaxKind.JsxSpreadAttribute:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function jsxTagNameExpressionValidation() {
    // todo: some way to validate this
}
/* istanbul ignore next */
function objectLiteralElementLikeAliasValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.PropertyAssignment:
        case typescript_1.SyntaxKind.ShorthandPropertyAssignment:
        case typescript_1.SyntaxKind.SpreadAssignment:
        case typescript_1.SyntaxKind.MethodDeclaration:
        case typescript_1.SyntaxKind.SetAccessor:
        case typescript_1.SyntaxKind.GetAccessor:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function caseOrDefaultClauseValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.CaseClause:
        case typescript_1.SyntaxKind.DefaultClause:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function moduleReferenceValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.Identifier:
        case typescript_1.SyntaxKind.QualifiedName:
        case typescript_1.SyntaxKind.ExternalModuleReference:
            return;
        default:
            var ensureNever = value;
    }
}
/* istanbul ignore next */
function typeElementTypes() {
    // todo: some way to validate this
}
/* istanbul ignore next */
function templateLiteralValidation() {
    var value = null;
    switch (value.kind) {
        case typescript_1.SyntaxKind.TemplateExpression:
        case typescript_1.SyntaxKind.NoSubstitutionTemplateLiteral:
            return;
        default:
            var ensureNever = value;
    }
}
