// ts-simple-ast/dist-scripts/src/tests/compiler/base/questionTokenableNodeTests.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var testHelpers_1 = require("../testHelpers");
describe("QuestionTokenableNode", function () {
    function getInfoWithFirstPropertyFromText(text) {
        var result = testHelpers_1.getInfoFromText(text);
        return __assign({}, result, { firstProperty: result.firstChild.getInstanceProperties()[0] });
    }
    describe("hasQuestionToken", function () {
        function doTest(text, value) {
            var firstProperty = getInfoWithFirstPropertyFromText(text).firstProperty;
            chai_1.expect(firstProperty.hasQuestionToken()).to.equal(value);
        }
        it("should have a question token when has one", function () {
            doTest("class MyClass { prop?: string; }", true);
        });
        it("should not have a question token when not has one", function () {
            doTest("class MyClass { prop: string; }", false);
        });
    });
    describe("getQuestionTokenNode", function () {
        it("should be get the question token node", function () {
            var firstProperty = getInfoWithFirstPropertyFromText("class MyClass { prop?: string; }").firstProperty;
            chai_1.expect(firstProperty.getQuestionTokenNode().getText()).to.equal("?");
        });
        it("should be undefined when not optional", function () {
            var firstProperty = getInfoWithFirstPropertyFromText("class MyClass { prop: string;} ").firstProperty;
            chai_1.expect(firstProperty.getQuestionTokenNode()).to.be.undefined;
        });
    });
    describe("getQuestionTokenNodeOrThrow", function () {
        it("should be get the question token node", function () {
            var firstProperty = getInfoWithFirstPropertyFromText("class MyClass {\nprop?: string;}\n").firstProperty;
            chai_1.expect(firstProperty.getQuestionTokenNodeOrThrow().getText()).to.equal("?");
        });
        it("should throw when not optional", function () {
            var firstProperty = getInfoWithFirstPropertyFromText("class MyClass {\nprop: string;}\n").firstProperty;
            chai_1.expect(function () { return firstProperty.getQuestionTokenNodeOrThrow(); }).to.throw();
        });
    });
    describe("setHasQuestionToken", function () {
        function doTest(startText, value, expected) {
            var _a = getInfoWithFirstPropertyFromText(startText), firstProperty = _a.firstProperty, sourceFile = _a.sourceFile;
            firstProperty.setHasQuestionToken(value);
            chai_1.expect(sourceFile.getFullText()).to.be.equal(expected);
        }
        it("should be set as optional when not optional", function () {
            doTest("class MyClass { prop: string; }", true, "class MyClass { prop?: string; }");
        });
        it("should be set as optional when has an exclamation token", function () {
            doTest("class MyClass { prop!: string; }", true, "class MyClass { prop?: string; }");
        });
        it("should be set as not optional when optional", function () {
            doTest("class MyClass { prop?: string; }", false, "class MyClass { prop: string; }");
        });
        it("should do nothing when setting to same value", function () {
            doTest("class MyClass { prop: string; }", false, "class MyClass { prop: string; }");
        });
    });
    describe("fill", function () {
        function doTest(startCode, structure, expectedCode) {
            var _a = getInfoWithFirstPropertyFromText(startCode), firstProperty = _a.firstProperty, sourceFile = _a.sourceFile;
            firstProperty.fill(structure);
            chai_1.expect(sourceFile.getText()).to.equal(expectedCode);
        }
        it("should not modify when not set and structure empty", function () {
            doTest("class Identifier { prop: string; }", {}, "class Identifier { prop: string; }");
        });
        it("should not modify when set and structure empty", function () {
            doTest("class Identifier { prop?: string; }", {}, "class Identifier { prop?: string; }");
        });
        it("should modify when setting true", function () {
            doTest("class Identifier { prop: string; }", { hasQuestionToken: true }, "class Identifier { prop?: string; }");
        });
    });
});
